### PR TITLE
NEXUS-8187: Maven CMA Group w/ Content Attributes

### PR DESCRIPTION
**Pull request by cstamas**
Maven CMA Group w/ Content Attributes

Issue
https://issues.sonatype.org/browse/NEXUS-8187

### CODE REVIEW

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/Content.java [hunk 1]
```diff
@@ -18,33 +18,45 @@
...
  *
  * @since 3.0
  */
 public class Content
     implements Payload
 {
-  private final Payload payload;
+  public static final String CONTENT_LAST_MODIFIED = "lastModified";
+
+  public static final String CONTENT_ETAG = "etag";
```
Comment (jdillon) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/Content.java]: why are these defined here but not used by this class?

Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/Content.java]: Simply put, I could not find more proper place for them. 

They are attribute keys of a content, content itself does not use these, true. But if you look _where_ they are used (proxy facet support, maven proxy facet, maven facet, maven group facet...) the use is spread over multiple facets not quite tied logically to each other, to call out one of them as "the proper" place for these (and these are not maven specific at all). For now am collecting them here, and we can see where to move it.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/group/GroupHandler.java [hunk 2]
```diff
@@ -69,40 +72,82 @@ public String toString() {
...
+      final Response response = view.dispatch(request);
+      if (response.getStatus().getCode() == HttpStatus.OK) {
+        return response;
+      }
+    }
+    return HttpResponses.notFound();
+  }
+
+  /**
+   * Returns all responses from all members as a linked map, where order is processing order (actually group member
```
Comment (kellyrob99) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/group/GroupHandler.java]: can skip the 'actually' bit here; if the order is group member order just say so ;)

File: plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenGroupFacet.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+    final Metadata mergedMetadata = metadataMerger.merge(metadatas.toArray(new MetadataEnvelope[metadatas.size()]));
+    if (mergedMetadata == null) {
+      return null;
+    }
+    final ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+    new MetadataXpp3Writer().write(byteArrayOutputStream, mergedMetadata);
+    // Metadata is rather small, so let's do it like this
+    final Iterable<HashAlgorithm> hashAlgorithms = Iterables.transform(Arrays.asList(HashType.values()),
+        new Function<HashType, HashAlgorithm>()
```
Comment (kellyrob99) [inline plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenGroupFacet.java]: this function could be shared(also defined in MavenFacetImpl)

File: plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenGroupFacet.java [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+          @Override
+          public HashAlgorithm apply(final HashType input) {
+            return input.getHashAlgorithm();
+          }
+        });
+    final Map<HashAlgorithm, HashCode> hashCodes = Hashes.hash(hashAlgorithms,
+        new ByteArrayInputStream(byteArrayOutputStream.toByteArray()));
+    final Content content = new Content(
+        new BytesPayload(
+            byteArrayOutputStream.toByteArray(),
```
Comment (kellyrob99) [inline plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenGroupFacet.java]: minor, but no need to create the byte array twice, assign to var should save us some memory

File: plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenGroupFacet.java [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    content.getAttributes().set(Content.CONTENT_LAST_MODIFIED, DateTime.now());
+    content.getAttributes().set(Content.CONTENT_ETAG, "{SHA1{" + hashCodes.get(HashAlgorithm.SHA1).toString() + "}}");
+    content.getAttributes().set(Content.CONTENT_HASH_CODES_MAP, hashCodes);
+
+    // cache the metadata and the hashes
+    mavenFacet.put(mavenPath, content);
+    for (HashType hashType : HashType.values()) {
+      final HashCode hashCode = hashCodes.get(hashType.getHashAlgorithm());
+      if (hashCode != null) {
+        mavenFacet.put(mavenPath.hash(hashType), new StringPayload(hashCode.toString(), "text/plain"));
```
Comment (kellyrob99) [inline plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenGroupFacet.java]: Personal preference, we should use constants for mime types if possible: https://github.com/sonatype/nexus-oss/blob/master/components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentTypes.java#L20-20

**Comment by kellyrob99** [comment]
+1

File: plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenGroupFacet.java (hunk 5 above)
Comment (cstamas) [inline plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenGroupFacet.java]: ack, fixing

File: plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenGroupFacet.java (hunk 3 above)
Comment (cstamas) [inline plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenGroupFacet.java]: ack, I thought about that too, fixing

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/group/GroupHandler.java (hunk 2 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/group/GroupHandler.java]: ack, fixing

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/partial/PartialFetchHandler.java [hunk 6]
```diff
@@ -64,8 +64,8 @@ public Response handle(@Nonnull final Context context) throws Exception {
       return response;
     }
 
-    if (!response.getStatus().isSuccessful()) {
-      // We don't interfere with failure responses
+    if (!response.getStatus().isSuccessful() || response.getStatus().getCode() == HttpStatus.PARTIAL_CONTENT) {
```
Comment (mrprescott) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/partial/PartialFetchHandler.java]: Now that I look them over, I'm not sure it makes sense to partially fetch anything other than 200 - wdyt?

Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/partial/PartialFetchHandler.java]: True, only 200 OK payload responses should be processed

File: plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenFacetImpl.java [hunk 7]
```diff
@@ -112,22 +113,39 @@
 
   /**
    * List of hash algorithms always applied to payloads in blob store, with first in the list being the "primary" hash.
+   * In case of Maven, this list is driven by {@link HashType}.
+   *
+   * @see HashType
    */
-  private static final List<HashAlgorithm> HASH_ALGORITHMS = ImmutableList.of(SHA1, MD5);
+  private static final List<HashAlgorithm> HASH_ALGORITHMS = ImmutableList.copyOf(
```
Comment (mrprescott) [inline plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenFacetImpl.java]: Heh, maybe just a constant in HashType ?
    public final static List<HashAlgorithm> ALGORITHMS = ImmutableList
        .of(SHA1.getHashAlgorithm(), MD5.getHashAlgorithm());

File: plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenFacetImpl.java [hunk 8]
```diff
@@ -112,22 +113,39 @@
...
 
+  private MavenPathParser mavenPathParser;
+
   private VersionPolicy versionPolicy;
 
   private ChecksumPolicy checksumPolicy;
 
   @Inject
-  public MavenFacetImpl(final MimeSupport mimeSupport) {
+  public MavenFacetImpl(final MimeSupport mimeSupport, final Map<String, MavenPathParser> mavenPathParsers) {
```
Comment (mrprescott) [inline plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenFacetImpl.java]: This is here to provide eventual support for a maven1 path parser?

Comment (cstamas) [inline plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenFacetImpl.java]: Yes

File: plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenFacetImpl.java (hunk 7 above)
Comment (cstamas) [inline plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenFacetImpl.java]: Ack, fixing

File: plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenGroupFacet.java [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+        final Metadata metadata = reader.read(inputStream);
+        metadatas.add(new MetadataEnvelope(origin, metadata));
+      }
+      catch (XmlPullParserException e) {
+        // skip it, log it
+        log.info("Unparseable repository metadata: {}", origin, e);
+      }
+    }
+
+    final Metadata mergedMetadata = metadataMerger.merge(metadatas.toArray(new MetadataEnvelope[metadatas.size()]));
```
Comment (mrprescott) [inline plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenGroupFacet.java]: Makes me wonder if merge() should take Iterable<Metadata> not be varargs.

File: plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenGroupFacet.java [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+      }
+    }
+
+    if (metadataContents.isEmpty()) {
+      return null;
+    }
+    return doMergeAndCacheMetadata(mavenPath, metadataContents);
+  }
+
+  private Content doMergeAndCacheMetadata(final MavenPath mavenPath,
```
Comment (mrprescott) [inline plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenGroupFacet.java]: I think a 50-line method with the word 'and' in its name is a clear sign that it's doing too much in a single method. :P

File: plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/Maven2GroupRecipe.groovy [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+import org.sonatype.nexus.repository.view.handlers.TimingHandler
+
+import static org.sonatype.nexus.repository.http.HttpHandlers.notFound
+
+/**
+ * Raw group repository recipe.
+ *
+ * @since 3.0
+ */
+@Named(Maven2GroupRecipe.NAME)
```
Comment (mrprescott) [inline plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/Maven2GroupRecipe.groovy]: @Singleton I think

**Comment by mrprescott** [comment]
+1 with minor comments

File: plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/Maven2GroupRecipe.groovy (hunk 11 above)
Comment (cstamas) [inline plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/Maven2GroupRecipe.groovy]: Ack, fixing

File: plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenGroupFacet.java (hunk 9 above)
Comment (cstamas) [inline plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenGroupFacet.java]: Ack, fixed. This really looks wrong as is

File: plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenGroupFacet.java (hunk 10 above)
Comment (cstamas) [inline plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenGroupFacet.java]: Method split with 45862c0ce308435ac5b306f5da6c48e537059cdd

**Comment by cstamas** [comment]
Merged/squashed as 8cbd14f1bd0f10d0e388a7addd7b45c507ded63f

**Pull request closed without merging**